### DISCUSSION TITLE
What is the best way to get all valid paths for a website?

### DISCUSSION THREAD

**Discussion by ScottPierce** [vikejs/vike#1037] [Discussion] [answered]
Using SSR, is there a way to know or calculate a complete list of all the valid page paths for the website at runtime?

**Comment by brillout** [comment, marked as answer]
What's your use case?

I believe you should be able to access `pageContext._pageRoutes` but note that this is internal and there is therefore no guarentee that it won't break at any (minor) version update.

(FYI some sponsors use internal APIs while their use case is included in VPS's CI ensuring that no release breaks their app without us knowing it beforehand.)

**Reply by ScottPierce** [reply]
It's difficult to explain, but I'll give it a shot. We're trying to setup something with this plugin that runs multiple servers off of a single code base running your plugin. If a user makes a change, or updates our data, we need to quickly be able to re-deploy a page, set of pages, or their entire site.

We have another server (the deploy server) that's responsible for calling into the SSR server (the rendering server), then uploading it into an environment. We have dynamic pages, and static pages. Dynamic pages are stored in a database, so the deploy server is more than capable of getting and understanding what those are, but it's not capable of knowing what the static pages are for a given site. The rendering server, ultimately is the only one with this data, so it needs to be able to tell the deploy server all possible routes at runtime so that it can render each of them.

**Reply by ScottPierce** [reply]
FYI `pageContext._pageRoutes` is exactly what we were looking for. Thank you.

**Reply by brillout** [reply]
That's a neat use case. (These kind of use cases and, in general, the need for more flexibility is one of the reasons why I created vite-plugin-ssr.)

I really think sponsoring can make sense for you. Especially given that you use vite-plugin-ssr USPs; I'm not sure you could achieve your use case with other exsiting tools.

Looking forward to further support and solidify your use case.

**Reply by StanleyDing** [reply]
@brillout The other day I'm looking at the pageContext on the client and find this `_pageRoutes`. I am wondering why it is exposed to the client, and if I don't want the client to know all the paths in the app (e.g. admin pages), is there any way that I can exclude these paths?

**Reply by brillout** [reply]
@StanleyDing That's inherent to Client Routing

**Reply by StanleyDing** [reply]
@brillout Thank you for the clarification! However, the `_pageConfigs`, which contains all pageId, is still exposed in `pageContext` even if my page is configured to use server routing.

**Reply by brillout** [reply]
@StanleyDing Feel free to create a feature request elaborating your use case. Please don't write in unrelated discussions like this one, thanks.
